Fix tpm joining protobuf oneof

strideynet opened the pull request:
These protos aren't in use yet - I made a small mistake in the original PR by using a oneof. As far as I can tell, there's some weirdness around `oneof` and GoGoProto/jsonpb/protojson, and I'd prefer to avoid that footgun if I can.

strideynet committed:
Fix tpm joining protobuf oneof

codingllama commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1276,25 +1276,24 @@ message ProvisionTokenSpecV2 {
+    // has submitted an EKCert, and the public key in the EKCert will be used
+    // for this check.
+    // Example: d4b45864d9d6fabfc568d74f26c35ababde2105337d7af9a6605e1c56c891aa6
+    string EKPubHash = 4 [(gogoproto.jsontag) = "ekpub_hash,omitempty"];
```
This is alright, but what is the problem with oneofs that you are trying to avoid? I don't think oneofs are always a good choice, but I can't say I've had issues with it.

strideynet replied:
Hm - perhaps I'm overly worried. I saw some comments referring to how different JSONpb impls handle oneof's and a quick inspection revealed that basically no other Teleport resources leverage oneof's.

In addition, when I thought this over, I realised that maybe there's a valid intention in specifying both - e.g I want to limit joining to both a specific EKCert serial and an EKPub. I can't see this being particularly useful day to day, but it could be more interesting if we expanded TPM joining to support things like LDevID certs in future which are installed by the Owner rather than the Platform.

codingllama replied:
> I realised that maybe there's a valid intention in specifying both

Yep, that's the real caveat of oneofs, realizing that you do want those fields together sometimes.

codingllama commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -1276,25 +1276,24 @@ message ProvisionTokenSpecV2 {
 // ProvisionTokenSpecV2
 message ProvisionTokenSpecV2TPM {
   message Rule {
+    reserved 2, 3;
```
Linters will want you to reserve/change the names too.

strideynet replied:
Hmm - I don't really want to change the name.

codingllama replied:
I've got bad news for you...

strideynet replied:
:(

strideynet replied:
I'll come up with a better but slightly similar name 😂

codingllama replied:
Brilliant  😂

strideynet committed:
See if Buf will allow me to reuse the indexes outside of oneof

strideynet committed:
Regen protos

strideynet committed:
Tactical renaming

strideynet committed:
Merge branch 'master' into strideynet/fix-tpm-join-protos

strideynet committed:
Update CRDs

strideynet committed:
Met push
rge branch 'strideynet/fix-tpm-join-protos' of github.com:gravitational/teleport into strideynet/fix-tpm-join-protos

strideynet commented:
@hugoShaka
